lib/spack/spack/test/spec_semantics.py

```diff
@@ -9,280 +9,348 @@
+            ("libelf debug=2", "libelf foo=1", "libelf debug=2 foo=1"),
+            ("libelf debug=2", "libelf debug=2 foo=1", "libelf debug=2 foo=1"),
+            ("libelf+debug", "libelf~foo", "libelf+debug~foo"),
+            ("libelf+debug", "libelf+debug~foo", "libelf+debug~foo"),
+            ("libelf++debug", "libelf+debug+foo", "libelf++debug++foo"),
+            ("libelf debug==2", "libelf foo=1", "libelf debug==2 foo==1"),
+            ("libelf debug==2", "libelf debug=2 foo=1", "libelf debug==2 foo==1"),
+            ("libelf++debug", "libelf++debug~foo", "libelf++debug~~foo"),
+            ("libelf foo=bar,baz", "libelf foo=*", "libelf foo=bar,baz"),
+            ("libelf foo=*", "libelf foo=bar,baz", "libelf foo=bar,baz"),
+            (
+                'multivalue-variant foo="bar"',
+                'multivalue-variant foo="baz"',
+                'multivalue-variant foo="bar,baz"',
+            ),
+            (
+                'multivalue-variant foo="bar,barbaz"',
+                'multivalue-variant foo="baz"',
+                'multivalue-variant foo="bar,baz,barbaz"',
+            ),
+            # Flags
+            ("mpich ", 'mpich cppflags="-O3"', 'mpich cppflags="-O3"'),
+            (
+                'mpich cppflags="-O3 -Wall"',
+                'mpich cppflags="-O3 -Wall"',
+                'mpich cppflags="-O3 -Wall"',
+            ),
+            ('mpich cppflags=="-O3"', 'mpich cppflags=="-O3"', 'mpich cppflags=="-O3"'),
+            (
+                'libelf cflags="-O3"',
+                'libelf cppflags="-Wall"',
+                'libelf cflags="-O3" cppflags="-Wall"',
+            ),
+            (
+                'libelf cflags="-O3"',
+                'libelf cppflags=="-Wall"',
+                'libelf cflags="-O3" cppflags=="-Wall"',
+            ),
+            (
+                'libelf cflags=="-O3"',
+                'libelf cppflags=="-Wall"',
+                'libelf cflags=="-O3" cppflags=="-Wall"',
+            ),
+            (
+                'libelf cflags="-O3"',
+                'libelf cflags="-O3" cppflags="-Wall"',
+                'libelf cflags="-O3" cppflags="-Wall"',
+            ),
+        ],
+    )
+    def test_abstract_specs_can_constrain_each_other(self, lhs, rhs, expected):
+        """Test that lhs and rhs intersect with each other, and that they can be constrained
+        with each other. Also check that the constrained result match the expected spec.
+        """
+        lhs, rhs, expected = Spec(lhs), Spec(rhs), Spec(expected)
 
-def check_constrain(expected, spec, constraint):
-    exp = Spec(expected)
-    spec = Spec(spec)
-    constraint = Spec(constraint)
-    spec.constrain(constraint)
-    assert exp == spec
+        assert lhs.intersects(rhs)
+        assert rhs.intersects(lhs)
 
+        c1, c2 = lhs.copy(), rhs.copy()
+        c1.constrain(rhs)
+        c2.constrain(lhs)
+        assert c1 == c2
+        assert c1 == expected
```

Comment:
Here I always try to check all the invariant we expect for each of the new parametrized tests.